src/blob.rs

```diff
@@ -348,9 +348,9 @@ impl<'a> BlobObject<'a> {
         };
 
         let clean = sanitize_filename::sanitize_with_options(name, opts);
-        let mut iter = clean.splitn(2, '.');
-        let stem: String = iter.next().unwrap_or_default().chars().take(64).collect();
+        let mut iter = clean.rsplitn(2, '.');
```

Comment:
Cant you *not* compromise here? no random bits, no nothing. Take filename, then decide accept/reject. When it is reject then *maybe* generate stem+"_counter" so that foo.tar.gz becomes foo_1.tar.gz etc (just counting up until there is no collision). Same when the filename is too long. 

But then, the important part: don't use that as is but ask the user for editing/confirmation of the generated name. in 99% thats prolly just a OK from the user, while for the remaining 1% its important that the user has a chance to correct it. 

Think about filenames like  "foobar-checksum.tar.gz" becoming truncated to "foobar-che.tar.gz" which would destroy some important metadata that was passed in the name itself.

Reply:
> Take filename, then decide accept/reject. When it is reject then maybe generate stem+"_counter" so that foo.tar.gz becomes foo_1.tar.gz etc (just counting up until there is no collision).

I don't understand what you mean with "mabye", we will always need a filename to save the attachment as.

Counting up would be possible, but take longer sometimes (because we would have to look whether the filename already exists) and message recieving/sending should go as quickly as possible.

> But then, the important part: don't use that as is but ask the user for editing/confirmation of the generated name. in 99% thats prolly just a OK from the user, while for the remaining 1% its important that the user has a chance to correct it.

Anything involving new UI needs coordination amongst lots of folks and lots of work overall; I'd guess something like 30 hours of work overall here. We unfortunately don't have the capacities to do this.

Reply:
> but when this lossy conversion results in a string that cant be accepted (As much I see only the length limit would be a problem) then reject that on import.

reject? We need a name for the file, and we can't just say "sorry, you can't send this file, we don't like the filename"

> > Counting up would be possible, but take longer sometimes (because we would have to look whether the filename already exists) and message recieving/sending should go as quickly as possible.
> 
> I am aware of that, but for any real world case this shouldn't be a problem. Even counting up into some hundreds to thousands may be some 'slight' performance impact but would be rare enough and acceptable since this is a pretty cold code path. In some other code iirc I used a u16 and error out once that overflows.

Yeah, maybe... But I don't really see why it would be worth it, I mean, most users don't look at the filename that much anyway (and WhatsApp even often completely scrambles the filename).

Reply:
Ideally no filename gets rejected, but unfortunately that is not the real world, at least not when you want to be portable, whats a legal filename on the sending end (aka already exists) might be invalid or misinterpreted on the receiving end.

I opened the other ticket about garbled extension just because a similar a problem, sending from DC-Desktop a "Foo.Bar.pdf" and a friend received it at "Foo.Bar.pd" on Android, which there is a pretty bad user experience because it wont open up in a pdf viewer and you can't *easly* rename a file on android (save it, open filemanager, if any is installed, hope it has permissions to access/rename it, figure out how to rename...).

Filenames are not (always) utf-8 strings, they add some non portable restrictions about characters usable and usable length. Even when you define it as rust Path, that's not portable. For a sender there is never the trouble because the filename already exists on the filesystem, but eventually you have to do some decisions/handle error on the receiver side at least.

Still I would prefer when the Sender could straighten this out in a sensible way (Ideally by getting a acceptable filename presented to be confirmed). Reason is that this fixed the filename 'once for all' in case the file is send to multiple recipients or forwarded by one of them. As well the sender is aware of the new Filename when referring to it. There could be still an error creating a file on the receiver side, but that's another story.

To the rescue: Once you are aware of the problems it is reasonably easy to sanitize filenames in most cases. They can be transmitted as utf-8, invalid characters could be substituted (underscore or U+FFFD REPLACEMENT CHARACTER) some characters should be not permitted  (\/:...) for portability reasons (rejected, removed or replaced). For acceptable lengths one can just make something up that covers almost any modern OS (say 128 'characters', not bytes, not utf-8). The receiver is then (anyway) responsible to recreate the filename, in proper encoding.

The hard decisions are what to do when a filename is too long:

In my earlier response i shown that it is undecidable for corner cases what constitutes a extension (when the filename contains dots).  "how.to.send.pdf.gz" is this a .pdf.gz or a compressed textfile about how to send pdfs? There is always a fair chance that *any* heuristic you encode is wrong.

Based on this knowledge you can deduce that you can not decide safely how to shorten a filename past the first dot, which leaves little room ins such (very rare) cases. So unenviable you have either to reject some rare filenames or offer the user a way to make it compliant.